### ISSUE TITLE
Update configuration file in non persistent enviroments

### ISSUE THREAD

**Issue by alexlopezcifuentes** [ultralytics/ultralytics#21261]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

## Context

We have a running training pipeline working with Ultralytics. Data sources for datasets are properly configured so that everything runs smoothly. We have configured our `dataset_dir` to be a different path than the one set by default in:

https://github.com/ultralytics/ultralytics/blob/78770420f90a99d4d9bc799750269a52032898bd/ultralytics/utils/__init__.py#L1017

## Issue

The issue we are facing occurs when running the training pipeline in non-persistent environments, i.e., a Databricks cluster. Databricks clusters, just for some prior context, are basically fresh every time you start one. This means that the configuration file `.config` disappears and every time a cluster is started, it is created from zero.

We have seen that there is a way of updating this config from Python using:

```python
from ultralytics import settings
# Update a setting
settings.update({"dataset_dir": "/path/to/data"})
```

The issue occurs when this `settings.update()` method does not seem to properly update the internal YOLO variables, because if we do:

```python
from ultralytics import settings
# Update a setting
settings.update({"dataset_dir": "/path/to/data"})
# Setup model
model = YOLO(cfg.model.name)
# Train the model
model.train()
```
it is still looking for data in the old config path. The run will fail and if we run the script again it will now properly work.

**Questions**:

- Is this the expected behavior of the `settings.update()` function?
- Should it not actually update all the internal variables depending on the configuration file?

### Additional

****

**Comment by UltralyticsAssistant** [comment]
👋 Hello @alexlopezcifuentes, thank you for reaching out and for providing detailed context about your training pipeline and environment! 🚀 This is an automated response to help get you started—an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for guidance on configuration and environment setup. If you're using Python, check out the [Python Usage Examples](https://docs.ultralytics.com/usage/python/); for command-line workflows, see our [CLI Usage Examples](https://docs.ultralytics.com/usage/cli/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the behavior in your non-persistent environment. This will help us investigate and resolve your issue more efficiently.

If you have additional questions about custom training or environment configuration, please include as much detail as possible, such as code snippets, dataset samples, and logs. For best results, make sure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, join our [Discord](https://discord.com/invite/ultralytics) 🎧
- For deeper discussions, visit [Discourse](https://community.ultralytics.com/)
- To engage with others, check out our [Subreddit](https://www.reddit.com/r/Ultralytics/) 🤗

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to confirm the issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for reporting this! Your input helps us improve Ultralytics HUB and the YOLO ecosystem.

**Comment by Y-T-G** [comment]
I guess it's because the values are cached during initialization, so it doesn't get updated after initialization.

https://github.com/ultralytics/ultralytics/blob/78770420f90a99d4d9bc799750269a52032898bd/ultralytics/utils/__init__.py#L1578

You could update it using the CLI `yolo settings` command before launching your training script. That way it has the right `datasets_dir` before initialization.

**Comment by Spritea** [comment]
I use the code below to change the settings of `runs_dir`. I think it should also work for other settings, like `datasets_dir`.
```python
from ultralytics.utils import SettingsManager
settings = SettingsManager()
settings.update({"runs_dir": "/path/to/runs_dir"})
```

**Comment by glenn-jocher** [comment]
@alexlopezcifuentes This is expected behavior due to how settings are cached during initialization. The issue occurs because settings are loaded once when the module is first imported, so updating them afterward doesn't affect already-cached values.

For non-persistent environments like Databricks, update your settings before any YOLO imports or model initialization:

```python
from ultralytics import settings
settings.update({"datasets_dir": "/path/to/data"})
# Now import and use YOLO
from ultralytics import YOLO
model = YOLO(cfg.model.name)
model.train()
```

Alternatively, you can use the environment variable `YOLO_CONFIG_DIR` to specify a custom config directory path, which gets checked during initialization as mentioned in the [get_user_config_dir function](https://docs.ultralytics.com/reference/utils/__init__/#ultralytics.utils.get_user_config_dir).
